libc/sem.cc

```diff
@@ -76,3 +105,79 @@ int sem_trywait(sem_t* s)
         return libc_error(EAGAIN);
     return 0;
 }
+
+static std::unordered_map<std::string, indirect_semaphore> named_semaphores;
+static mutex named_semaphores_mutex;
+
+OSV_LIBC_API
+sem_t *sem_open(const char *name, int oflag){
+    SCOPE_LOCK(named_semaphores_mutex);
+    std::unordered_map<std::string, indirect_semaphore>::iterator iter = 
+                                named_semaphores.find(std::string(name));
+    if(iter != named_semaphores.end()){
+        iter->second.add_reference();
+        return reinterpret_cast<sem_t*>(&(iter->second));
+    }
+    
+    errno = ENOENT;
+    return SEM_FAILED;
+}
+
+OSV_LIBC_API
+sem_t *sem_open(const char *name, int oflag, mode_t mode, unsigned int value){
+    SCOPE_LOCK(named_semaphores_mutex);
+    std::unordered_map<std::string, indirect_semaphore>::iterator iter = 
+                                named_semaphores.find(std::string(name));
+    
+    if(oflag & O_CREAT){
+        if(iter != named_semaphores.end()){
+            if(oflag & O_EXCL){
+                errno = EEXIST;
+                return SEM_FAILED;
+            }
+
+            iter->second.add_reference();
+            return reinterpret_cast<sem_t*>(&(iter->second));
+        }
+        
+        if(value > SEM_VALUE_MAX){
+            errno = EINVAL;
+            return SEM_FAILED;
+        }
+        /*
+         * TODO: mode bits. Do we actually want to worry about permissions in a
+         * unikernel? Would also need to check permissons when using open and unlink
+        */
+        named_semaphores.emplace(std::string(name), indirect_semaphore(value, 1, true)); 
+        return reinterpret_cast<sem_t *>(&named_semaphores[std::string(name)]);
+    }
+
+    errno = ENOENT;
+    return SEM_FAILED;
+}
+
+OSV_LIBC_API
+int sem_unlink(const char *name){
+    SCOPE_LOCK(named_semaphores_mutex);
+    std::unordered_map<std::string, indirect_semaphore>::iterator iter = 
+                                named_semaphores.find(std::string(name));
+    if(iter != named_semaphores.end()){
+        iter->second.unlink();
```

Comment:
Since `sem_unlink` takes a name, rather than a `sem_t *`, I don't think it can called on an unnamed semaphore and so we don't have to worry about trying to call `unlink` on a type that doesn't have that method. I believe that when a name is passed that is already unlinked (i.e. not in the named_semaphores map), it should fail with `ENOENT` as the error number. I took your advice and created some tests in a Linux environment and the results seem consistent with the explanations above.